spec/p2p/p2p-top-level.md

```diff
@@ -0,0 +1,310 @@
+  aspects, system assumptions and temporal
+  logic specifications.
+
+  - [Incentives](#incentives): how faulty full nodes may benefit from
+    misbehaving and how correct full nodes benefit from cooperating.
+  
+  - [Computational Model](#Computational-Model):
+      timing and correctness assumptions.
+
+  - [Distributed Problem Statement](#Distributed-Problem-Statement):
+      temporal properties that formalize safety and liveness
+      properties in the distributed setting.
+
+- [Part IV](#part-iv---Protocol):
+  Specification of the protocols.
+
+     - [Definitions](#Definitions): Describes inputs, outputs,
+       variables used by the protocol, auxiliary functions
+
+     - [Protocol](#core-verification): gives an outline of the solution,
+       and details of the functions used (with preconditions,
+       postconditions, error conditions).
+
+     - [Liveness Scenarios](#liveness-scenarios): when the light
+       client makes progress depends heavily on the changes in the
+       validator sets of the blockchain. We discuss some typical scenarios.
+
+- [Part V](#part-v---supporting-the-ibc-relayer): Additional
+  discussions and analysis
+
+
+In this document we quite extensively use tags in order to be able to
+reference assumptions, invariants, etc. in future communication. In
+these tags we frequently use the following short forms:
+
+- TMBC: Tendermint blockchain
+- SEQ: for sequential specifications
+- LIVE: liveness
+- SAFE: safety
+- FUNC: function
+- INV: invariant
+- A: assumption
+
+--->
+
+# Part I - A Tendermint node 
+
+TODO: 
+- perhaps we should survey here what we think are the expectations of the reactors from p2p? 
+    - consensus might need for liveness that neighborhood is stable
+    - mempool might just need that a transaction can reach each validator within a reasonable amount of time (this might be achievable if at no point in time the current graph is connected, but, e.g., something along the lines that the union of the graphs over some period in time is connected.)
+- What does p2p expect from the reactors? (don't falsely report bad nodes; this puts requirements on the reactors and perhaps/likely also on the application running on top of ABCI)
+
+## Context of this document
+
+> mention other components and or specifications that are relevant for this
+spec. Possible interactions, possible use cases, etc. 
+
+> should give the reader the understanding in what environment this component
+will be used. 
+
+TODO:
+- Reactor API
+- Network? How do we communicate with other nodes
+- Discuss that validators run special set-up, and manage their own neighborhood (hide behind sentry nodes).
+   - As a result: the distributed system is composed of
+       - (correct) nodes that follow the protocol described here
+       - (potentially) adversarial nodes whose behavior deviates to harm the system
+       - (correct) nodes that don't follow the protocol to shield themselves but behave in a "nice way"
+
```

Comment:
Would it be worth touching on "connectedness" in this section? Maybe some definitions that give an idea of how the number connections that a node keeps to peers, together with the "amount of bad peers" (or probability of a peer to be bad) has an impact on the global properties we want to achieve (e.g. a message reaches all full nodes eventually)

Reply:
Yes, you are right. I think there are basic implicit assumptions/requirements in the current solution, e.g., between any two correct validators there is a path consisting only of correct full nodes. (Most likely even more paths). This might be a requirement that the p2p needs to provide to the reactors, and it can only do so, if the underlying network has even stronger properties.

Reply:
Moreover, I hardly can see this property being ensure, it is problematically achieved.

Reply:
You are right @sergio-mena, in that these properties cannot be guaranteed, so maybe we shouldn't try. I mean, the P2P should only guarantee a best effort guarantee of conectedness, and then we call out that if by some external forces other guarantees are made, such as the existence of a GST and a maximum number of byzantine nodes, then and stronger properties can be derived, like eventual delivery of messages from some correct processes to some correct processes. This way we modularize the guarantees.